Added Collaborator, Contributor and Repository.collaborators(affiliation)

AndreasBackx opened the pull request:
See #730 for some background information. This still fails 1 test because the cassette `Repository_collaborators.json` needs to be updated with the permissions as it somehow does not contain that information. The GitHub API does say that it should be there, but the recording is from 2014 so perhaps it wasn't returned back then.

Note that I have currently put the named parameter `affiliation` at the front of the arguments in `Repository.collaborators`. I did this to be in line with the other similar iterators. This can be moved to the back, but I assume some backwards-incompatible changes are fine for 1.0.0?

This uses separate classes for the specific User object uses as I understood the idea was from #670.

AndreasBackx commented:
@sigmavirus24 any chance this gets looked at?

sigmavirus24 commented:
@AndreasBackx I'd be happy to give this a look. Would you mind rebasing this to merge cleanly?

sigmavirus24 commented on github3/repos/users.py:
```diff
@@ -0,0 +1,33 @@
+from ..users import _User
```
Is there reason to not just put these classes in `github3.users`?

AndreasBackx replied:
No reason at all, preference perhaps. It'll get changed.

AndreasBackx replied:
Oh wait, now that I think about it. The reason I put it in `github3.repos.users` instead of `github3.repos.users` is because I assumed that was implied in the issue in the issue #670. As a collaborator's endpoint is `/repos/:owner/:repo/collaborators`. They belong to the repo so I thought it would be good to put those in `github3.repos.users` instead. Regardless though, tell me what you think is best.

sigmavirus24 commented on github3/repos/repo.py:
```diff
@@ -524,17 +524,24 @@ def code_frequency(self, number=-1, etag=None):
         """
         url = self._build_url('collaborators', base_url=self._api)
-        return self._iter(int(number), url, users.ShortUser, etag=etag)
+        affiliations = {'outside', 'direct', 'all'}
```
Eh. I've waffled on this before. I think we do something similar in other places. I've wondered about having a `github3.constants` or (`github3._constants`) but it never felt important enough, honestly.

sigmavirus24 replied:
I also toyed with some decorators that would do strong validation of the parameters and what's passed in to the function but that turned out to be way too much work.

AndreasBackx replied:
We can create an enum `Affiliation` perhaps with uppercase choices for each: `Affiliation.OUTSIDE`. You can then do `list(Affiliation)` for a list of all of them. Maybe something along those lines could be adopted for other stuff as well. I don't really mind here. Tell me what to do for the constants here, and it'll get done.

sigmavirus24 replied:
I don't think you need to tackle this @AndreasBackx. I think this is a murky corner for a library that supports Py2 and Py3. If we were Py3 only, I'd go whole-hog on using an Enum. Relying on the `enum34` backport, however, has its own issues, I've found, and isn't worth the headache for end users right now.

AndreasBackx replied:
Ah yes, I've been working exclusively in Python 3 so I've been forgetting about the Python 2 support as of late. Are there any plans to drop Python 2 support in the future?

sigmavirus24 commented on github3/repos/repo.py:
```diff
@@ -524,17 +524,24 @@ def code_frequency(self, number=-1, etag=None):
+                "'direct', or 'all' (default 'all')."
+            )
+        params = {'affiliation': affiliation}
+        return self._iter(int(number), url, Collaborator, params, etag=etag)
```
I think as a rule we've switched to importing *modules* and not classes, so these shouldn't be the exception.

sigmavirus24 commented on github3/repos/repo.py:
```diff
@@ -524,17 +524,24 @@ def code_frequency(self, number=-1, etag=None):
+        affiliations = {'outside', 'direct', 'all'}
+        if affiliation not in affiliations:
+            raise ValueError(
+                "Invalid affiliation parameter passed, must be 'outside', "
```
I wonder if we should format the given affiliation into this to make it easier for debugging. Perhaps:

```py
("Invalid affiliation value {!r} parameter passed, must be 'outside',"
 "'direct', or 'all' (defaults to 'all').").format(affiliation)
```

sigmavirus24 commented on tests/unit/test_repos_repo.py:
```diff
@@ -1042,10 +1042,29 @@ def test_collaborators(self):
+        )
+
+    def test_collaborators_valid_affiliation(self):
+        """
```
I know we're not great at this elsewhere in the tests, but I'm hoping to get our doc-strings cleaned up soon. I tried to do so in the code itself. I'm going to target the tests next and it'd be good to do less work :wink: 

Could you substitute something like

```
"""Verify we can list collaborator with a valid affiliation.

Let's ensure that our affiliation validation works.
"""

AndreasBackx replied:
Probably a mistake on my end. I usually follow the same principle.

sigmavirus24 commented:
Gave this a review pass as well. There's no rush to fix it up. This looks really close to perfect. Thanks @AndreasBackx for putting the work into doing this. :tada:

AndreasBackx commented:
Alright, I've looked over the comments and I'll fix what needs to be fixed and I've replied to your questions. I'll try to get to it this week.

sigmavirus24 commented:
Oh, and regarding how `Contributor` and `Collaborator` turn into a `User`, I think the answer lies in the examples of how `ShortUser` becomes a `User`(i.e., setting `_refresh_to`)

AndreasBackx commented:
I'm rebasing right now and see that Contributor has the attribute `contributions`, but it seems to be documented as `contributions_count`. Is the documentation wrong or should the attribute be renamed?

A User has got:

```python
self.followers_count = user['followers']
```

And the Contributor has got:

```python
class Contributor(_User):
    """Object for the specialized representation of a contributor.

    When retrieving a repository's contributors, GitHub returns the same
    information as a :class:`~github3.users.ShortUser` with an additional
    attribute:

    .. versionadded:: 1.0.0

        This class was added in version 1.0.0

    .. attribute:: contributions_count

        The number of contributions a contributor has made to the repository

    """

    class_name = 'Contributor'

    def _update_attributes(self, contributor):
        super(Contributor, self)._update_attributes(contributor)
        self.contributions = contributor['contributions']
```

Should it become:

```python
self.contributions_count = contributor['contributions']
```

I thought it'd post it here and include it in the PR perhaps, it seems a bit silly to make a new PR/issue just for this.

AndreasBackx commented:
I've rebased the PR following your feedback. I haven't changed `Contributor.contributions` to `Contributor.contributions_count` like I mentioned in my previous comment.

I've added the ability to refresh the `Collaborator` as well like @sigmavirus24 mentioned and added it to the docs. I assumed that this will be in 1.0.3 so I've used that version in the version changes/additions.

I had to update the cassette `Repository_collaborators`, but because I don't have access to this repository, which is used in the tests. I changed it temporarily to my PR's develop branch, ran the tests and changed the username from AndreasBackx to sigmavirus24. It shouldn't make a difference for the tests that are performed using it or in the future I guess.

Let me know if I missed something.

sigmavirus24 commented:
It should be `contributions_count` that's what I intended it to be at least :)

sigmavirus24 commented:
Let me try closing and re-opening this to trigger builds. Travis was [broken](https://twitter.com/sigmavirus24/status/982994790785716224) so let's see if this fixes AppVeyor too.

AndreasBackx commented:
I've reverted `Repository_collaborators.json` to its original state as the base64 strings in it contain the URLs to my repository. It needs to be regenerated by someone who has push access to this repository.

I've also found another "bug" as you will in the tox config:

```ini
envlist = py{27,34,35,36,py},py{27,34,36}-flake8,docstrings
```
The envlist mentions a `py36-flake8` environment, but that one does not exist so it falls back to `py36` which then essentially gets run twice when running `tox`.

sigmavirus24 commented:
I re-recorded the cassette and pushed it to your branch. @AndreasBackx can you send another PR to add that environment to our tox.ini?

sigmavirus24 commented:
I pushed some extra changes here as I noticed some things I didn't quite like about the docstrings.

sigmavirus24 commented:
:tada: Thanks @AndreasBackx!

AndreasBackx commented:
Regarding the environment change in `tox.ini` do I need to add it or do I need to change the py34-flake8 to py36-flake8? I doubt that we need 2 of those environments for Python 3?

sigmavirus24 commented:
You can change `py34` to `py36` for Flake8. Just make sure you remove the `py34` from the envlist.

And don't worry about checking in the old cassette. It's all sorted :smile:

AndreasBackx commented:
Will do.

I know, I saw the diff which is when I realised that I had accidentally checked it in.
